### PR TITLE
Reduce matrix to upper hessenberg form with #hessenberg

### PR DESCRIPTION
**Pull request by v0dro**
tests for hess

hessen wip

wip

hess done

### CODE REVIEW

File: lib/nmatrix/math.rb [hunk 1]
```diff
@@ -227,6 +227,27 @@ def factorize_lu with_permutation_matrix=nil
...
+    raise ShapeError, "Trying to reduce non 2D matrix to hessenberg form" if 
+      shape.size != 2
+    raise ShapeError, "Trying to reduce non-square matrix to hessenberg form" if 
+      shape[0] != shape[1]
+    raise StorageTypeError, "Matrix must be dense" if stype != :dense
+    raise TypeError, "Works with float matrices only" unless 
+      [:float64,:float32].include?(dtype)
+
+    a = self.clone
+    __hessenberg__(a)
```
Comment (cjfuller) [inline lib/nmatrix/math.rb]: Might also be nice to have a `hessenberg!` that does this without the clone too.

**Comment by cjfuller** [comment]
Two general comments:

1. It's generally a good idea if you're moving around unrelated code (even if it makes sense) to do it in a separate pull request that only moves the code, so that it's easier to tell exactly what's necessary for the PR's particular feature.  (I spent an embarrassing amount of time trying to figure out what was going on with some of the unrelated methods in `ruby_nmatrix.c`.)

2. The C implementation of the algorithm is taken directly from the referenced source.  There's no license given for the code in that source and I can't find any license on that entire website.  The only thing I see is a "copyright, all rights reserved" notice, which would mean we can't use it without express permission (much like Numerical Recipes).  Not really sure what to do here, though, since it's a fairly small piece of code and many implementations of the same algorithm are going to look kind of similar.  Thoughts @mohawkjohn ?

**Comment by mohawkjohn** [comment]
I'm not seeing the copyright notice anywhere. Unless there's a clear permission in the license, we can't use the code.

**Comment by v0dro** [comment]
Should I write to the author and ask him for permission?
If he says yes, I can add his mail to a gist and post the link in the code.

And sorry for `ruby_nmatrix.c`, I'll keep that in mind henceforth.

**Comment by mohawkjohn** [comment]
That would be a good approach. Yes, you'd want to include his name in the comments.

**Comment by v0dro** [comment]
One question about copyrights - The website says copyright 2004. Doesn't that mean that its not applicable in 2015?

**Comment by mohawkjohn** [comment]
No, it just means that's the last time the copyright statement was updated. You'd have to ask a lawyer for the effect, but I think it's minimal.

**Comment by v0dro** [comment]
@cjfuller @mohawkjohn I've just received permission from the author (which I've forwarded to you both).

Can this be merged now?

**Comment by cjfuller** [comment]
Please add a note to the comments you have there re: the source saying "included by permission of the author" and then we can merge.

**Comment by v0dro** [comment]
@cjfuller updated

**Pull request merged**